pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionActioner.java

```diff
@@ -243,4 +257,70 @@ private String getDownloadPackagePath(FunctionMetaData functionMetaData, int ins
                 },
                 File.separatorChar);
     }
+
+    public static boolean isFunctionCodeBuiltin(FunctionDetailsOrBuilder functionDetails) {
+        if (functionDetails.hasSource()) {
+            SourceSpec sourceSpec = functionDetails.getSource();
+            if (!StringUtils.isEmpty(sourceSpec.getBuiltin())) {
+                return true;
+            }
+        }
+
+        if (functionDetails.hasSink()) {
+            SinkSpec sinkSpec = functionDetails.getSink();
+            if (!StringUtils.isEmpty(sinkSpec.getBuiltin())) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private File getBuiltinArchive(FunctionDetails.Builder functionDetails) throws IOException {
+        if (functionDetails.hasSource()) {
+            SourceSpec sourceSpec = functionDetails.getSource();
+            if (!StringUtils.isEmpty(sourceSpec.getBuiltin())) {
+                File archive = connectorsManager.getSourceArchive(sourceSpec.getBuiltin()).toFile();
+                String sourceClass = ConnectorUtils.getConnectorDefinition(archive.toString()).getSourceClass();
+                SourceSpec.Builder builder = SourceSpec.newBuilder(functionDetails.getSource());
+                builder.setClassName(sourceClass);
+                functionDetails.setSource(builder);
+
+                fillSourceSinkTypeClass(functionDetails, archive, sourceClass);
+                return archive;
+            }
+        }
+
+        if (functionDetails.hasSink()) {
+            SinkSpec sinkSpec = functionDetails.getSink();
+            if (!StringUtils.isEmpty(sinkSpec.getBuiltin())) {
+                File archive = connectorsManager.getSinkArchive(sinkSpec.getBuiltin()).toFile();
```

Comment:
quick question: if connector also requires external jar eg: KinesisSink requires authentication-jar which can be present into classpath of function-worker process. In that case, when worker loads function, will function be able to access that authentication jar?
I think it should work, as right now we keep auth-jar in worker-classpath and it can be accessed by function if we load function in `ThreadRunTime` but might need to think about `ProcessRuntime`.

Reply:
Where is the authentication jar coming from? Could that be included in the NAR archive?

Reply:
KinesisSink has an option to provide a classname of a credential-generation-provider using which it can generate aws-credential and it loads that provider dynamically at runtime (same as pulsar-broker loads authentication-provider dynamically). so, if user uses `built-in` kinesis-connector then credential-generation provider jar can't be part of the NAR archive. so, right now, we keep this provider-jar into function-worker classpath and `ThreadRuntime` can access it while initializing kinesisSink.